man/rebase-helper.1
.TH REBASE-HELPER 1 2016-07-31 "" "Linux User's Manual"
.SH NAME
rebase-helper \- helps you to rebase package to the latest upstream version

.SH SYNOPSIS
.B rebase-helper
[\fIOPTION\fR]... \fISOURCES\fR

.SH DESCRIPTION
\fBrebase-helper\fP is the tool which helps you to rebase package
to the latest upstream version defined by SOURCES argument

Mandatory arguments to long options are mandatory for short options too.
.TP
\fB\-b\fR, \fB\-\-build-only\fR
build SRPM and RPMs

.TP
\fB\-p\fR, \fB\-\-patch-only\fR
apply patches to new upstream version

.TP
\fB\-\-buildtool\fR=\fIBUILDTOOL\fR
tool for building old and new sources.
Supported [mock (default), rpmbuild, koji, copr].

.TP
\fB\-\-pkgcomparetool\fR=\fIPKGCOMPARETOOL\fR
select a tool which compares both RPM packages.
Supported [pkgdiff, rpmdiff, abipkgdiff, csmock].

.TP
\fB\-\-outputtool\fR=\fIOUTPUTTOOL\fR
select a tool for showing information from rebase-helper process.
Supported [text, json].

.TP
\fB\-\-not\-download\-sources\fr
suppress download sources from web

.TP
\fB\-c\fR, \fB\-\-continue\fR
select the choice if you want to continue with rebase previously interrupted

.TP
\fB\-v\fR
output is more verbose

.TP
\fB\-w\fR, \fB\-\-keep\-workspace\fR
select the choice if you want to keep workspace directory after finishing

.TP
\fB\-\-build\--retries\fR=\fIBUILD_RETRIES\fR
specifies number of retries rebase-helper will do in case of failed build.
If build fails, rebase-helper analyzes the build log, tries to fix the issue if possible and retries the build.

.SH AUTHOR
Written by Petr Hracek and Tomas Hozza


ChangeLog
News in version 0.8.0:
- Added support for JSON output format
- Added support for COPR build tool
- Renamed 'fedpkg' build tool to 'koji' to make it more clear
- Downloading of files is now done only using standard Python library and not using PyCURL
- Added support for passing arbitrary extra arguments to local builders (mock, rpmbuild) with '--builder-options'.
- Added new option '--build-retries' allows the user to specify number of build retries (by default 2)
- Added support for csmock check tool
- Many bug fixes and code clean up

News in version 0.6.0:
- calls patches and git operation if exists in %prep section
- support libabipkgdiff
- Several fixes
- replace calling yum with dnf

News in version 0.5.0:
- git rebase method is used instead of patch command
- several decoding issues were fixed
- DiffHelper class is not needed
- Several PEP8 and W1202 fixes
- continue option implemented for git rebase
- added several tests
- Class for operation with Git repositories

News in version 0.4.0:
- we now handle extra versions like 'b1', 'rc1', ...
- improved version extraction from archive name
- added Build Log analyzer to detect unpackaged files or other issues
- Added Bash completion
- logging rebase-helper output to rebase-helper-results directory
- rewrite SpecFile class

News in version 0.3.1:
- new build class
- build-only option
- installing build dependencies in case of rpmbuildtool
- more tests
- RebaseHelperError class for catching exceptions
- Several fixes

News in version 0.3.0:
- pkgdiff class for comparing RPM files
- test suite for archive class and spec file

News in version 0.2.0:
- diff_helper for comparing two tarballs
- applying patches to tarballs
- patch_helper

News in version 0.1.0:
- first draft of the project
- Initial classes
- CLI interface


rebase-helper.bash
# Bash completition for rebase-helper tool
# Copyright (C) 2013-2014 Red Hat, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# he Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# Author: Tomas Hozza <thozza@redhat.com>

# fill LOCAL_ARCHIVES variable with supported archive names
_local_archives()
{
    # supported archive extensions
    extensions=".tar.gz .tgz .tar.xz .tar.bz2 .zip .gem"
    LOCAL_ARCHIVES=()
    for ext in $extensions; do
        LOCAL_ARCHIVES=("${LOCAL_ARCHIVES[@]}" $( ls | grep $ext\$ ))
    done
}

# fill --*tool arguments into ARGUMENTS variable
_complete_tool_arguments()
{
    buildtools="mock rpmbuild koji copr"
    pkgcomparetools="pkgdiff rpmdiff abipkgdiff csmock"
    outputtools="text json"

    if [ $# != 1 ]; then
        ARGUMENTS=()
    else
        cmd="$( echo $1 | sed -E s'/--(.*)/\1/'g )s"
        ARGUMENTS=${!cmd}
    fi
}

_rebase-helper()
{
    local cur prev opts

    COMPREPLY=()
    cur="${COMP_WORDS[COMP_CWORD]}"
    prev="${COMP_WORDS[COMP_CWORD-1]}"

    # basic commands
    opts="-h --help \
          -v --verbose \
          -p --patch-only \
          -b --build-only \
          --buildtool \
          --pkgcomparetool \
          --outputtool \
          -w --keep-workspace \
          --not-download-sources \
          -c --continue \
          --non-interactive \
          --comparepkgs-only \
          --builds-nowait \
          --build-tasks \
          --results-dir \
          --build-retries \
          --builder-options"

    # complete arguments of some commands
    case "${prev}" in
        --*tool)
            _complete_tool_arguments ${prev}
            COMPREPLY=( $(compgen -W "${ARGUMENTS}" -- ${cur}) )
            return 0
            ;;
        *)
            ;;
    esac

    _local_archives
    # complete options as well supported archives in current directory
    COMPLETE="${LOCAL_ARCHIVES[@]} ${opts[@]}"
    COMPREPLY=( $(compgen -W "${COMPLETE}" -- ${cur}) )
    return 0
}

complete -F _rebase-helper rebase-helper
